source/convection_diffusion/assemble_rhs.cc

```diff
@@ -8,6 +8,187 @@
 namespace RMHD
 {
 
+namespace
+{
+
+
+template <int dim>
+void compute_source_term
+(Function<dim>* const  ptr,
+ const std::vector<Point<dim>> &quadrature_points,
+ const std::vector<double>     &gamma,
+ const double                   previous_time,
+ const double                   current_time,
+ const double                   next_time,
+ std::vector<double>           &source_term)
+{
+  AssertDimension(source_term.size(), quadrature_points.size());
+
+  std::vector<double> old_old_source_term_values(quadrature_points.size());
+  ptr->set_time(previous_time);
+  ptr->value_list(quadrature_points,
+                  old_old_source_term_values);
+
+  std::vector<double> old_source_term_values(quadrature_points.size());
+  ptr->set_time(current_time);
+  ptr->value_list(quadrature_points,
+                  old_source_term_values);
+
+  std::vector<double> source_term_values(quadrature_points.size());
+  ptr->set_time(next_time);
+  ptr->value_list(quadrature_points,
+                  source_term_values);
+
+  // Loop over quadrature points
+  for (std::size_t q=0; q<quadrature_points.size(); ++q)
+    source_term[q] =
+      (gamma[0] * source_term_values[q] +
+       gamma[1] * old_source_term_values[q] +
+       gamma[2] * old_old_source_term_values[q]);
+}
```

Comment:
While this indeed improves on the readability of the assembly, is it really efficient? In every call the `std::vector` instances are created and destroyed. By keeping them in the scratch data, their values are only updated.